Add database

KlochkovDimka opened the pull request:
Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
Добавлены контролеры GenreController.java, MpaController.java
Обновлены модели
Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

KlochkovDimka committed:
feat:
1. Сформирована структура базы данных H2 в файле schema.sql, дополнительные данные в файле data.sql
2. Созданы Dao классы: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Добавлены контролеры GenreController.java, MpaController.java
4. Обновлены модели
5. Добавлены новые исключения NotExistGenreException.java, NotExistMpaBadRequestException.java, NotExistMpaException.java
6. Добавлены тесты для взаимодействия с базой данных

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -12,26 +12,25 @@
  * Film.
  */
-@Data
+@Setter
+@Getter
+@ToString
```
А зачем разбивать аннотацию `@Data` на три?

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -1,10 +1,21 @@
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Positive;
+
+@Service
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
```
Можно использовать аннотацию `@Data`

LexLippi commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,98 @@
+DROP TABLE IF EXISTS PUBLIC.LIKES_FILMS;
+DROP TABLE IF EXISTS PUBLIC.FRIENDSHIP;
+DROP TABLE IF EXISTS PUBLIC.STATUS_FRIENDSHIP;
+DROP TABLE IF EXISTS PUBLIC.USERS;
+DROP TABLE IF EXISTS PUBLIC.GENRES_FILMS;
+DROP TABLE IF EXISTS PUBLIC.GENRES;
+DROP TABLE IF EXISTS PUBLIC.FILMS;
+DROP TABLE IF EXISTS PUBLIC.MPA;
```
Вот от этого необходимо избавиться, так как ни один пользователь не захочет пользоваться приложением, которое удаляет данные при каждом обновлении

LexLippi commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,98 @@
+);
+CREATE UNIQUE INDEX IF NOT EXISTS PRIMARY_KEY_7 ON PUBLIC.GENRES (ID);
+--------------------------------------------------------ЖАНРЫ-ФИЛЬМА--------------------------------------------------------
+CREATE TABLE IF NOT EXISTS PUBLIC.GENRES_FILMS (
```
Для правильной работы таблиц необходимо, чтобы в них был первичный ключ

LexLippi commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,98 @@
+--------------------------------------------------------ДРУЖБА--------------------------------------------------------
+CREATE TABLE IF NOT EXISTS PUBLIC.FRIENDSHIP (
+  USER_ID INTEGER NOT NULL AUTO_INCREMENT,
+  FRIENDS_ID INTEGER,
+  STATUS_ID INTEGER,
```
Могут быть `null`?

LexLippi commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,28 @@
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (1, 'Комедия');
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (2, 'Драма');
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (3, 'Мультфильм');
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (4, 'Триллер');
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (5, 'Документальный');
+INSERT INTO PUBLIC.GENRES (ID, NAME)
+VALUES (6, 'Боевик');
+
+INSERT INTO PUBLIC.MPA (MPA_ID, NAME)
+VALUES (1, 'G');
+INSERT INTO PUBLIC.MPA (MPA_ID, NAME)
+VALUES (2, 'PG');
+INSERT INTO PUBLIC.MPA (MPA_ID, NAME)
+VALUES (3, 'PG-13');
+INSERT INTO PUBLIC.MPA (MPA_ID, NAME)
+VALUES (4, 'R');
+INSERT INTO PUBLIC.MPA (MPA_ID, NAME)
+VALUES (5, 'NC-17');
+
+INSERT INTO PUBLIC.STATUS_FRIENDSHIP (STATUS_ID, NAME)
+VALUES (1, 'TRUE');
+INSERT INTO PUBLIC.STATUS_FRIENDSHIP (STATUS_ID, NAME)
+VALUES (2, 'FALSE');
```
Лучше использовать оператор `merge`

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/users/userStorage/UserDbStorage.java:
```diff
@@ -0,0 +1,129 @@
+                    userRow.getDate("birthday").toLocalDate());
+            user.setId(userRow.getLong("id"));
+            return Optional.of(user);
+        } else {
```
В `else` нет необходимости, так как мы делаем `return` в основной ветке условного оператора. Эквивалентно будет написать так:

```java
if (userRow.next()) {
            User user = new User(
                    userRow.getInt("id"),
                    userRow.getString("email"),
                    userRow.getString("login"),
                    userRow.getString("name"),
                    userRow.getDate("birthday").toLocalDate());
            user.setId(userRow.getLong("id"));
            return Optional.of(user);
}
throw new NotExistUserException("User not found");
```

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/users/userStorage/UserDbStorage.java:
```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+    private Optional<User> newUser() {
```
Зачем в таком случае возвращать `Optional`, если он всегда заполнен?

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/users/userStorage/UserDbStorage.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private Optional<User> newUser() {
+        SqlRowSet userRow = jdbcTemplate.queryForRowSet("SELECT * FROM USERS u  ORDER BY u.ID DESC LIMIT 1;");
```
Зачем `ORDER BY`, если далее делаем `LIMIT 1` это не имеет смысла

KlochkovDimka replied:
Привет, этот метод возвращал только что сохраненного пользователя в базу данных. Здесь используется ORDER BY DESP с LIMIT 1, чтобы получить последнее добавление в БД

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/users/userStorage/UserDbStorage.java:
```diff
@@ -0,0 +1,129 @@
+                    userRow.getDate("birthday").toLocalDate());
+            user.setId(userRow.getLong("id"));
+            return Optional.of(user);
+        } else {
```
`else` не нужен, аналогично тому, что писал ниже

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/users/userStorage/InMemoryUserStorage.java:
```diff
@@ -1,33 +1,34 @@
+import java.util.*;
 
 @Component
+@Qualifier("InMemoryUserStorage")
```
Можно больше не поддерживать этот компонент

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/mpa/impl/MpaStorageDaoImpl.java:
```diff
@@ -0,0 +1,47 @@
+
+@Component
+@Slf4j
+public class MpaStorageDaoImpl implements MpaStorageDao {
```
Замечания аналогичные тем, что писал ниже

LexLippi reviewed (changes requested):
Привет!

Неплохая работа!

Есть некоторое количество замечаний, которые необходимо исправить!

Желаю удачи!

KlochkovDimka committed:
feat:
1. Добавлены первичные ключи для корректной работы таблиц schema.sql, изменения в data.sql
2. Доработки в Dao классах: FriendshipDbStorageDao.java, GenresFilmsStorageDao.java, LikesFilmsStorageDao.java, MpaStorageDao.java и имплементации их функций: FilmDbStorage.java, FriendshipDbStorageDaoImpl.java, GenresFilmsStorageDaoImpl.java, LikesFilmsStorageDaoImpl.java, MpaStorageDaoImpl.java, UserDbStorage.java
3. Изменения к классах-моделях связанные с аннотациями lombok

LexLippi commented:
Привет!

Хорошая работа!

Желаю удачи в следующих спринтах!
